### DISCUSSION TITLE
How to create rust functions that take a C function pointer as argument?

### DISCUSSION THREAD

**Discussion by ananis25** [PyO3/pyo3#2514] [Questions]
Hi - I am trying to write a python extension module using PyO3. The rust section orchestrates the computation while the actual function to execute is provided at runtime from python. 

For regular python functions, I can pass them like any other arg and PyO3 casts it to a `PyObject`.  
```rust
#[pyfunction]
fn rust_caller(py: Python, callback: PyObject, x: i64, y: i64) -> PyResult<PyObject> {
    println!("Orchestration logic");
    for n in 0..100 {
        callback.call(py, (x, y), None)
    }
}
```

However the computation is numerical, so I use the `numba` library to compile the `callback` function to native code. Though, calling the compiled routine through python would acquire and release the GIL each time. 

Numba also produces a C function pointer that can be passed to native libraries - [link](https://numba.readthedocs.io/en/stable/user/cfunc.html). This repository illustrates using it - [link](https://github.com/synapticarbors/rust-numba-test/blob/master/src/lib.rs).
Is it possible to create a rust function using PyO3 that takes an `extern C fn(...)` as an argument?

**Comment by davidhewitt** [comment]
Sorry for the very slow reply.

I don't think it will be possible to create a `#[pyfunction]` that will do this. Instead you should do the same as the example repository you link: create `unsafe extern "C" fn` and use `ctypes` to call it. Unfortunately, you should be aware of #1683 (hopefully we can fix this incompatibility eventually).
